docs/migration/v12.md

```diff
@@ -0,0 +1,277 @@
++       await findByRole('progressbar')
+    })
+})
+```
+
+#### Enzyme
+
+Enzyme is [considered dead](https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl): there is no official adapter for React 18 (or 17), so it's best to [migrate to React Testing Library](https://testing-library.com/docs/react-testing-library/migrate-from-enzyme/) for tests, if possible.
+
+If it's not currently possible to migrate, there is an [unofficial React 18 adapter](https://github.com/cfaester/enzyme-adapter-react-18) that can help in the mean time.
+
+## Troubleshooting & tips
+
+-   There is a bug in Vite: importing from a directory, e.g. `‘./app/‘` where we would assume to get `./app/index.[jt]s`, can incorrectly resolve to a file with a similar name to the directory, e.g. `./App.tsx`. I see this as a bug with Vite, but as a workaround, file and dir names can be changed so they don’t conflict. Noticed in the Maintenance app beta.
+-   If you’re using `identity-obj-proxy` in your test config, make sure you add it to the project’s dependencies. Previously, some apps got away with using the proxy without an explicitid dependency because the package is a dependency of `react-scripts`
+-   You may see some build warnings coming from the `mathjs` package. Upgrading to a recent version of `mathjs` will handle these warnings. Noticed in analytics app (the analytics library has a dependency on `mathjs`)
+
+### Summary of breaking changes
+
+1. Node 18 || 20+ is required
+2. JSX in files without .jsx or .tsx extensions is not supported by default
+3. React 18: Most significant testing changes are described above; the rest can be read about in the React [migration guide](https://react.dev/blog/2022/03/08/react-18-upgrade-guide)
+4. Flow types won’t work, although this is a motivator for extensible config
+5. Import aliases will likely break (also a motivator for extensible config)
+6. `global.variableName` is no longer supported; use `window.variableName` instead
+7. In order to make the dev server available on LAN, the `--host` flag needs to be passed to the start script
+8. Workers imported in the “webpack” way will need to be imported in a new way: [docs](https://vitejs.dev/guide/features.html#web-workers)
+9. App.jsx entrypoint is no longer quietly added if a plugin entrypoint is defined or for a defined but empty entrypoints object
+10. Plugins and apps are started together in the same process on the same port
+11. The `PORT` env var is removed
+12. `import * as MyClass from ‘my-module’` is now more restrictive, and may break in some cases. So far, changing the import to `import MyClass from ‘my-module’` has fixed it. DOMPurify is one example case, noticed in the Login App
+
+### Deprecations
+
+-   Environment variables on `process.env` will be dropped in future versions, in favor of variables on `import.meta.env`
+-   The default `direction` config value is `ltr`; in the future, it will be `auto`
+-   `--allowJsxInJs` will be removed
```

Comment:
wil this be removed in the underlying Vite CLI?

Reply:
Not sure which bullet point you mean haha, but here's an answer for each:
1. Env vars are under our control, but I'm actually rethinking this point, i.e. maybe we should keep variables on `process.env` -- Earlier, I thought `import.meta.env` was a pretty standard thing with ES6 and that we should migrate to that for env vars, but I think that's actually just a Vite pattern, so it might be "more standard" to keep env vars on `process.env`. If we do that, we can at least deprecate the `REACT_APP_` prefixes for the variable keys
2. (Probably obvious, but direction is also under our control)
3. For `--allowJsxInJs`, that's also under our control -- it's a flag that our own CLI reads, and adds a bunch of kinda complex Vite config in `makeViteConfig.mjs`. That additional config may or may not be supported in future versions of Vite, but I think we should remove _our_ `--allowJsxInJs` flag and config in the next `cli-app-scripts` version to make things simpler, standardized, and more performant